src/coreclr/src/vm/zapsig.cpp

```diff
@@ -241,10 +244,12 @@ BOOL ZapSig::GetSignatureForTypeHandle(TypeHandle      handle,
 
         // During IBC profiling this calls
         //     code:Module::TokenDefinitionHelper
+        // During MulticoreJit profiling this calls
+        //     code:MulticoreJitManager::TokenDefinitionHelper
         (*this->pfnTokenDefinition)(this->context.pModuleContext, pTypeHandleModule, index, &token);
 
         // ibcExternalType tokens are actually encoded as mdtTypeDef tokens in the signature
-        _ASSERT(TypeFromToken(token) == ibcExternalType);
+        _ASSERTE(this->context.externalTokens != ZapSig::MulticoreJitTokens && TypeFromToken(token) == ibcExternalType);
```

Comment:
I'm seeing this assertion fail because `externalTokens == MulticoreJitTokens` when the generic type parameter is a value type from a different assembly. Repro:

```c#
    private static void Main()
    {
        ProfileOptimization.SetProfileRoot(Environment.CurrentDirectory);
        ProfileOptimization.StartProfile("profile.mcj");

        Thread.Sleep(500);

        CallFoo();

        ProfileOptimization.StartProfile(null);
    }

    [MethodImpl(MethodImplOptions.NoInlining)]
    private static void CallFoo() => Foo(RegexOptions.IgnoreCase);

    [MethodImpl(MethodImplOptions.NoInlining)]
    private static void Foo<T>(T t)
    {
    }
```

I'm not familiar with the zapsig code to evaluate those changes or what might be wrong. It may be good to add some tests involving module-load dependencies as part of the type signature. The rest of the changes look good to me.

Reply:
It is my mistake. Actually, I want to pass if externaltokens are Multicorejittokens.
I updated zapsig encoding for multicorejit.
pfnTokenDefinition is not necessary for multicorejit. I removed. And I added some tests.